smart-contract-feature-development.md

```diff
@@ -0,0 +1,94 @@
+The deploy script should be able to parse some sort of config and optionally deploy the feature enabled version of the contract.
+
+### Handling Inheritance
+
+Inherit the contract that needs new functionality. It is very easy to add net new functions. If certain functions need to be modified,
+then the original contract may need to have the `virtual` keyword added in places. The `virtual` keyword should only be added in
+places where it must be added and nowhere in addition.
+
+### Handling Semver
+
+The parent contract should update its `version` getter to be a real function. The child contract should implement a `version` getter like so:
+
+```solidity
+function version() public override returns (string) {
+  return string.concat(super.version(), "+feature-name");
+}
+```
+
+The `feature-name` should be some identifier that uniquely identifies the feature. It could also be a hardfork name if the release is tied to a hard fork.
+
+### Handling Storage
+
+Handling storage becomes complex and there is no solution that handles all cases well. We generally want to ensure that the contracts
+are forwards compatible, so if `solc` is managing storage, the base contract should have spacers when we know that a feature contract
+is scheduled to be merged soon.
+
+Another solution is to use unstructured storage, meaning not letting `solc` manage storage and instead using custom storage slots.
+Assuming that digests and unique preimages are used for the storage slot keys, then this can work well for ensuring there are no
+storage slot collisions. This works well if there is no need for `solc` to manage the storage layout.
+
+If there are multiple feature contracts based on the same base contract being worked on in parallel, then there is complexity around
+layout management and it is up to the teams working on features to coordinate with each other if they risk mangling storage layouts.
+Using unstructured storage greatly reduces the coordination overhead here.
+
+More attention should be paid when the contracts are going to be deployed to a production network. It is less important to have the
+storage layouts 100% perfect when the contracts are going to be used on a transient network like a devnet. It is up to the developers
+to be mindful of their own upgrade paths as projects may have varying requirements.
+
+Given enough time, there will be storage layout issues. In the worst case, they can be fixed at upgrade time using the `StorageSetter`
+contract, this pattern is already used in production for `Initializable`.
+
+### Deploy Script
+
+The deploy script should be able to parse the deploy config in a standard way where the same config option enables the feature across the
+entire codebase. This means that a single config key/value pair should be able to have predeploys set in the L2 genesis, have the feature
+enabled contracts deployed to L1 and also tell op-node/op-geth to use that particular feature.
+
+### Merging Contracts
```

Comment:
I am a bit confused about what is meant by merging contracts. Is composition through inheritance only meant to be a temporary solution during the development phase, and then contracts will eventually be consolidated into one using feature flags to define behaviour?

Reply:
> Is composition through inheritance only meant to be a temporary solution during the development phase

Yes exactly this. We currently do not have a paradigm where we deploy different contracts to enable different features. This could be a design space we explore in the future, but I lean against it as we are already trying to handle versioning of a single set of contracts, see https://github.com/ethereum-optimism/design-docs/pull/21